File: Laboratório/LAB07/Largada.c

// Bibliotecas
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>



// Definição do tipo Corrida
typedef              	
  struct Corrida_struct  	
  {                   	
    int n_competidores;
    int	*inicial;
    int *final;	
  }                   	
Corrida_; 

typedef Corrida_* Corrida;

/*	Sets e gets 	*/

int getN (Corrida corrida) {
  return (corrida->n_competidores);
}

int get1Ini (Corrida corrida, int x) {
  return (corrida->inicial[x]);
}

int get1Fim (Corrida corrida, int x) {
  return (corrida->final[x]);
}

void setN (Corrida corrida, int n) {
  corrida->n_competidores = n;
}

void set1Ini (Corrida corrida, int x, int n) {
  corrida->inicial[n] = x;
}

void setIni (Corrida corrida, int *ini) {
  corrida->inicial = ini;
}

void set1Fim (Corrida corrida, int x, int n) {
  corrida->final[n] = x;
}

void setFim (Corrida corrida, int *fim){
  corrida->final = fim;
}

/*		*/

/*	Construtor	*/

Corrida newCorrida (int n, int *inicio, int *fim) {

Corrida novo;
novo = malloc ( sizeof (Corrida) );

setN (novo, n);
setIni (novo, inicio);
setFim (novo, fim);

}

/*		*/

/*	Método para obter o resultado de ultrapassagens	*/

int race_results (Corrida corrida) {

  //Dados
  int tamanho = getN (corrida);
  int ultrapassagens = 0;
  int x = 0, a = 0;

  while ( x < tamanho ) {

    a = x;
    // Achar onde ocorre o elemento 'x' do array de fim no de inicio
    while ( get1Ini(corrida, a) != get1Fim(corrida, x) ) {
      a++;
    }
    
    // "Método de inserção" até serem iguais
    while ( get1Ini(corrida, x) != get1Fim(corrida, x) ) {
      int temp = get1Ini(corrida, a);
      set1Ini ( corrida, get1Ini(corrida, a-1),   a );
      set1Ini ( corrida,                  temp, a-1 );
    
      a--;
      ultrapassagens++;
    }
  
    x++;
  }

  return ( ultrapassagens );
}

/*		*/

/*	Método main	*/
int main (void)
{

  // Variaveis e dados
  int n = 0, x = 0, y = 0;
  int *inicio;
  int *fim;
  int ultrapassagens;
  Corrida* competicoes = malloc ( 10 * sizeof(Corrida) );

  // Ler número de competidores e iniciar while
  while ( scanf("%d", &n) != EOF ) {

    //Iniciar arrays com base no número de competidores
    inicio = (int*) malloc ( n * sizeof(int) );
    fim = (int*) malloc ( n * sizeof(int) );

    // Ler arrays
    for (int i = 0; i < n; i++) {
	scanf( "%d", &inicio[i] );
    }
    for (int i = 0; i < n; i++) {
	scanf( "%d", &fim[i] );
    }

    //Iniciar classe corrida
    competicoes[x] = newCorrida (n, inicio, fim);

    x++;
  }

  for (y = 0; y < x; y++) {

    ultrapassagens = race_results(competicoes[y]);

    printf("%d\n", ultrapassagens);

  }

  return ( 0 );
}

/*		*/

File: Laboratório/LAB07/Largada.java

class Corrida {

  int n_competidores;
  int[] inicial;
  int[] final;

/* Construtores */
  Corrida () {

    this.n_competidores = 5;

    this.inicial = new int [this.getN()];
    this.final = new int [this.getN()];
  
  }

  Corrida(int n, int[] inicio, int[] fim) {

    this.setN(n);

    this.setIni(inicio);
    this.setFim(fim);

  }

  /* */

  /* Gets e sets */

  public int getN() {
    return (this.n_competidores);
  }

  public int get1Ini(int x) {
    return (this.inicial[x]);
  }

  public int get1Fim (int x) {
    return (this.final[x]);
  }

  public void setN(int n) {
    this.n_competidores = n;
  }

  public void setIni(int[] inicial) {
    this.inicial = new int[this.getN()];
    this.inicial = inicial;
  }

  public void setFim (int [] final) {
    this.final = new int [this.getN()];
    this.final = final;
  }
/* */

} // Fim da classe Corrida

class Largada {

  // Método para obter o número mínimo de ultrapassagens
  public int race_results (int [] inicio, int [] fim) {

  // Dados
  int tamanho = inicio.length;
  int ultrapasssagens = 0;
  int x = 0;
  int a = 0;
  
  while ( x < tamanho ) {
  
    a = x;
    // Achar onde ocorre o elemento 'x' do array de fim no de inicio
    while (inicio[a] != fim[x]) {
      a++;
    }
    
    // "Método de inserção" até serem iguais
    while (inicio[x] != fim[x]) {
      int temp = inicio [a];
      inicio [a] = inicio [a-1];
      inicio [a-1] = temp;
      
      a--;
      ultrapasssagens++;
    }
  
    x++;
  }


  return ( ultrapasssagens );
  }

  public void main (String[] args) {

    // Dados (Aplicar classe da corrida)
  
    int n = 0, x = 0, y = 0;
    String competidores_inicio;
    String competidores_fim;
    int [] inicio;
    int [] fim;
    int ultrapasssagens;
    Corrida[] competicoes = new Corrida[10];
  
    // Ler numero de competidores
    n = MyIO.readInt();
    while (n != EOF) {
  
      // Iniciar Strings e arrays com base no número de competidores
      inicio = new int [n];
      fim = new int [n];
    
      // Iniciar classe 'Corrida'
      competicoes[x] = new Corrida (n, inicio, fim);
  
      // Ler numero de competidores
      n = MyIO.readInt();

      x++;
    }

    while (y < x) {

      // Método de comparação
      ultrapassagens = race_results(inicio, fim);
      //ultrapassagens = race_results(competicao);
        
      // Print do resultado
      System.out.println(ultrapassagens);

      y++;
    }


  }

}
